### PR TITLE
Add smart column type detection and UI for setting those types

### PR DESCRIPTION
**Pull request by waxlamp**
This PR adds a long sought after feature: automatic type detection of CSV columns, and UI for updating them before uploading.

The way this works is to:
1. use `papaparse` to parse each line into a JS data object.
2. subject each value to a rudimentary type detector (see below).
3. use a simple heuristic to detect the closest type to suggest for each column
4. present a dropdown UI to let the user adjust the types before upload

The types currently recognized are:
- `label`: a string type that is meant to hold arbitrary data for creating textual labels in a visualization
- `category`: a type holding a limited number of distinct values, meant to drive visualization for an ordinal or categorical variable
- `number`: a quantitative value
- `date`: a value that looks like a date in some way

`label` is a fallback categorization; essentially any column that doesn't fall into the other types is assigned as a label. This works out naturally because CSV data, by definition, comes in the door as a set of strings. `label` is also suggested for fields named `_key`, `_from`, and `_to`, since those are special types in Multinet/ArangoDB, and need to actually be string values, even if they look like numbers, etc.

`category` is detected if the column contains 10 or fewer distinct values.

`number` is detected if every value in the column is coercible to JavaScript's `Number` type (using `!Number.isNaN(Number(x))` as the litmus test).

`date` is detected if every value in the column is coercible to DayJS's date wrapper type. However, `number` receives priority over date, because DayJS considers low numeric values to be valid dates, and in the majority of cases, such columns are actually intended as numbers.

All of the above is meant only for heuristic suggestion of the column types. The user gets final say by manipulating the resulting dropdowns.

This PR just offers a UI component for the feature; as things stand, the server would need to take the type information and do something with it prior to validation. The other option is to change how the CSV uploader works so that it accepts JSON lines instead of CSV, and use the type fields to properly convert the data before upload. This would still require the server to try to use the `datetime` module (or something more advanced, such as `arrow` or `pendulum` to attempt to convert the string values to dates, then settle them into the database in some canonical format, probably [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)).

TODO
- [x] `trim()` the `value` variable when testing for Numberness
- [x] properly clear out the dropdowns when the X is clicked in the file upload dialog

### CODE REVIEW

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [inline src/components/TableDialog.vue:170]: Why is this in computed? Nothing is changing here so I don't see a reason not to put it in `data`.

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [inline src/components/TableDialog.vue:219]: If you're only throwing an error here to satisfy TypeScript, and you're okay with betting that this is actually impossible, you can do the following:

```suggestion
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const entry = columnTypes.get(key)!;
```

**Comment by waxlamp** [comment]
@jtomeck, this is the start of a feature that we'll want your help in fixing up the design for later. No need to do anything now, besides maybe take a look at what I've done so you have an idea of where my head is on this feature.

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [inline src/components/TableDialog.vue:232]: It really doesn't matter but the only way for a string to be falsy is if it's empty, so you could do this
```suggestion
            if (value && !Number.isNaN(Number(value))) {
```

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [inline src/components/TableDialog.vue:254]: Why have this initial if statement? Won't this will always be a noop?

Comment (AlmightyYakob) [reply src/components/TableDialog.vue:254]: Also, what is `rec` supposed to mean? I think either a better variable name or a comment explaining this would help

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [inline src/components/TableDialog.vue:269]: I think we could move this entire function to a utils file or something, so that it's not cluttering this component so heavily. It could instead return `typeRecs`, at which point you'd perform the highlighted operations in this component.

**Review by AlmightyYakob — CHANGES_REQUESTED** [review]
Overall looks good, and I really like this implementation! Just a few suggestions

File: src/components/TableDialog.vue
Comment (JackWilb) [inline src/components/TableDialog.vue:250]: What about missing values here? That would mean that a column with all numbers and one missing value would be set to 'label' below.

File: src/components/TableDialog.vue
Comment (JackWilb) [reply src/components/TableDialog.vue:254]: I think `rec` is short for recommendation. I think that's reasonable. 

Also, the first statement has to exists, because the 4 boolean expressions above can all evaluate to false, which would cause issues. The default if we can't figure out the type should probably be label

File: src/components/TableDialog.vue
Comment (JackWilb) [reply src/components/TableDialog.vue:232]: That's not true (I think Roni mentioned this on slack). The string '0' is falsy

**Comment by JackWilb** [comment]
Also, @waxlamp are there libraries we could leverage for finding the types of the data instead of rolling our own? Or do we have a unique use case and types that we're searching for?

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [reply src/components/TableDialog.vue:254]: > I think `rec` is short for recommendation. I think that's reasonable.

I agree.

> Also, the first statement has to exists, because the 4 boolean expressions above can all evaluate to false, which would cause issues. The default if we can't figure out the type should probably be label

This is the section I commented on:

```
let rec: MultinetType = 'label';
if (isKey) {
  rec = 'label';
}
```

I'm saying the `if` statement there is redundant, as `'label'` is already the default value of `rec`.

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [reply src/components/TableDialog.vue:232]: I don't think that's true. If you evaluate the expression `!!'0'` in your browser console, the result is `true`. https://developer.mozilla.org/en-US/docs/Glossary/Falsy

Comment (JackWilb) [reply src/components/TableDialog.vue:232]: I stand corrected. I thought that `'0' == false` (which evaluates to true) implied falsy, but that's just js doing its thing with the double equals. Your change seems appropriate, @AlmightyYakob

File: src/components/TableDialog.vue
Comment (jtomeck) [inline src/components/TableDialog.vue:83]:
```suggestion
                v-model="columnType[field]"
                outlined
```

Comment (jtomeck) [reply src/components/TableDialog.vue:83]: Add outlined prop like all the other text fields in the app

**Review by jtomeck — COMMENTED** [review]
Besides the code changes I suggested, I wanted to ask. If you click the "x" to clear the file you selected, should those inputs disappear again?

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:170]: This is a habit of mine when I want to make sure the value being specified cannot be changed by accident. It's sort of the `const` of Vue components.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:219]: Even though mine is more verbose, I think I'll keep it. In this case, using the non-null assertion operator makes sense, but I do think it's somewhat harder to read, and the noise caused by having to suppress a rule seems to also go against legibility.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:232]: Nice catch. I think the exact confusion in this thread is the reason I like to be more explicit with this sort of test. I think it leaves the test more comprehensible by future readers as well.

However, you've made me realize there's a bug in the code: I should really be working with `value.trim()` instead of `value` to avoid having to worry about whitespace-only strings (which convert to `0` but are truthy).

I will add a commit to trim the value we're working with here.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:250]: In that case we'd have no choice but to recommend `label`--we can't coerce all the values to the target type.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:254]: `rec` was meant to be short for "recommendation", but `recommendation` seems too long for a variable name. If it's needed I can use `recommendation` or a different name for this variable.

I kept the `isKey` test for legibility. In pre-typescript days, I wouldn't have assigned any value to `rec` in its initializer, and then the if-elseif-else chain would be explicit and non-redundant. My choices with typescript are to (1) leave `rec` undefined and then type it as `MultinetType | undefined`; (2) cast the if-elseif-else chain as a function that gets called in-place (replacing the `rec =` parts with `return`) and assigned to `rec`; or (3) initialize the variable because I have to, and then use the expclicit if-elseif-else chain as I did.

(1) is a bad choice because it's going out of its way to do something simple.
(2) is a bad choice because it's confusing/ugly.
(3) is non-optimal because it looks redundant. HOWEVER, I think "optimizing" this code would be worse: it would leave it unclear that we deliberately select `label` for Multinet key fields.

This is another case of explicit code for the goal of comprehensibility. Let me know if this is ok or if you really want a different solution here.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:269]: Which entire function do you mean? Looking this over, it's a giant call to `Papa.parse()`, with the different callbacks being sort of related to each other in how this operates. I could see abstracting this whole call into a function outside the component definition but I don't see the point of moving just the type heuristic out, or of moving it to a global place like a utils file.

Let me know a bit more about what you're proposing and I can give you fuller thoughts on it.

**Comment by waxlamp** [comment]
> Also, @waxlamp are there libraries we could leverage for finding the types of the data instead of rolling our own? Or do we have a unique use case and types that we're searching for?

I am pretty sure there are, but you can see how simple the heuristics are that I chose to put into this PR. Keep in mind that the purpose is just to make a recommendation, and let the user make the final decision. We should do some research into type-wrangling libraries in case they are useful in the future, but I'm not sure there's a call for one in this PR.

File: src/components/TableDialog.vue
Comment (JackWilb) [reply src/components/TableDialog.vue:250]: We can't convert to a type with some missing values for blank rows? For example, having a number column with 5 missing values. This is standard in most data manipulation libraries/tools

Comment (waxlamp) [reply src/components/TableDialog.vue:250]: This is a good point. So far we've assumed that data landing in Multinet is "clean" in this sense (i.e., no missing values, etc.). I am not sure if that assumption is code-enforced on the server (@AlmightyYakob probably knows the validation routines best). But whether it is or not, this deserves a deeper discussion. Would you mind filing an issue so we can figure this out the right way?

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:83]: Thanks for this; updated.

**Comment by waxlamp** [comment]
> Besides the code changes I suggested, I wanted to ask. If you click the "x" to clear the file you selected, should those inputs disappear again?

They should; currently (as you've pointed out) they do not, and I also get some sort of failure in the console. Thanks for bringing this up; I will fix.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue]: See 21a337222bfea804dc0734e5b7bba5a6132d7643.

**Comment by waxlamp** [comment]
> > Besides the code changes I suggested, I wanted to ask. If you click the "x" to clear the file you selected, should those inputs disappear again?
> 
> They should; currently (as you've pointed out) they do not, and I also get some sort of failure in the console. Thanks for bringing this up; I will fix.

Fixed in b9a73979370f90080a26d8f6cb26dc17b300ec8b.

I also made the dropdowns a bit wider, and used the `dense` property on them (see f06ef6fed660eb352847798a388aeff04ef6e896).

**Review by jtomeck — APPROVED** [review]

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [reply src/components/TableDialog.vue:262]: I was missing the fact that the first `if` statement (that I claimed was redundant) has an effect, which is that it sets the priority of the checked conditions (`isKey` is the first checked condition, and if true will ensure that no other conditional statement overwrites `rec`). You could change the redundant assignment with the code below, but I'm also okay with keeping it as is. Up to you.
```
            let rec: MultinetType = 'label';
            if (!isKey) {
              if (category && !number && !date) {
                rec = 'category';
              } else if (number) {
                rec = 'number';
              } else if (date) {
                rec = 'date';
              }
            }
```

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [reply src/components/TableDialog.vue:172]: I think it's a bit of an anti-pattern, but I see the benefit of disallowing reassignment. I'm okay with this.

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [reply src/components/TableDialog.vue:277]: See #131 for my recommendation.

File: src/components/TableDialog.vue
Comment (AlmightyYakob) [reply src/components/TableDialog.vue:258]: I think most of the server-side validation is focused on tables matching each other (e.g. referenced keys existing), not so much column specifics. 

I think Jack's point is valid, one trivial way to address this would be to keep track of empty strings (kept track of in `TypeScore.empty`), and generate a type recommendation if that type count summed with the empty count equals the total. Example with number:

```
const number = entry.number + entry.empty === entry.total;
```

However this may be out of scope if we're just doing a first pass.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:172]: Since this is our test case for transitioning to composition API, it will become moot since in that context we will just use a `const` declaration. So I'll let this stand for now with the understand that it will be "fixed" in that manner.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:262]: I sort of like this approach but in the interest of keeping things moving, let's leave the code as is.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:277]: Approved #131. So I'll wait for you to merge it, and then we can conduct final reviews on this PR.

File: src/components/TableDialog.vue
Comment (waxlamp) [reply src/components/TableDialog.vue:258]: I like this idea but yeah it will need to wait until we decide on the semantics/practice of handling missing values in our API. I've linked this discussion into https://github.com/multinet-app/multinet-rfcs/issues/16.

**Review by AlmightyYakob — APPROVED** [review]

File: src/types.ts
Comment (JackWilb) [inline src/types.ts:3]: Looking at the data types [here](https://www.arangodb.com/docs/stable/aql/fundamentals-data-types.html) to see what types we could natively store in Arango, it looks like an easy one to add would be boolean. However, if we're not really planning to leverage the native types, then this point is irrelevant.

Another quick note, some of these data types in the docs above would be better supported by a JSON upload (for example Arrays and Objects), if we do decide to go for some native storage in Arango.

Comment (AlmightyYakob) [reply src/types.ts:3]: We have to be careful about this, because I think there's really two types flying around, which happen to intersect here. One is the actual data type that would be used to represent something in ArangoDB (JSON types: null, number, string, bool, etc.). The other is the "semantic" type, that is, how we choose to interpret the data in a column, regardless of it's actual data type. A good example of this might be if a a column is labeled as `date`, but each row entry could represent this with a number (unix timestamp) or a string (ISO 8601). Both would be valid ways to represent this data.

Maybe I'm incorrect about this, but regardless, I think more discussion is needed to flesh out the exact desired behavior. Thoughts @JackWilb @waxlamp?

Comment (JackWilb) [reply src/types.ts:3]: Right, thanks for explicitly spelling that out for us here. I agree that there are 2 different sets of types that we're talking about.

I guess the types that we use when storing data in Arango are actually irrelevant here since we can make a mapping from any type in the db to the types that we consider to be our "semantic" types.

Comment (waxlamp) [reply src/types.ts:3]: Good discussion. I like @AlmightyYakob's framing in terms of storage types and semantic types. And I like @JackWilb's observation that "semantic" types are more or less a hint to the vis app writer on what they may wish to do with the raw values coming from the database.

On the specific issue of a "bool" type: it is a good idea, but the difficulty is in execution. For instance, our heuristic could detect a column with only two values in it. Ordinarily, this would be marked as "categorical", but if the two values are strings `"true"` and `"false"` or `"yes"` and `"no"` then we might say it's a "bool" instead. Would we convert these values to literal bools in the database though? My gut says "no". For something like that, my intuition is that we leave it as "categorical" and let the client interpret the values. To have a first-class "bool" type we would really need to think about it (and we'd need a compelling use case first).

Good thoughts though!

Comment (waxlamp) [reply src/types.ts:3]: Oh, and on the topic of dates: one choice would be to do what BSON does, which is to serialize dates as this weird sort of structure:

```json
{
  "$date": {
    "value": "2020-10-16 16:41:21"
  }
}
```

(that's not precisely correct but it gets the idea across.)

If we sent up values like this from the client, that could serve as a decent signal that this is supposed to be a date.

I don't love this solution but our discussion made me think about BSON. So, just an idea to consider.

Comment (waxlamp) [reply src/types.ts:3]: > To have a first-class "bool" type we would really need to think about it (and we'd need a compelling use case first).

Just to follow up here, we do now have a compelling use case: UpSet. Let's talk about the boolean semantic type at our meeting on Wednesday.

**Pull request merged**
